cl_manycore/libraries/bsg_manycore_cuda.cpp

```diff
@@ -5,354 +5,1089 @@
+			if (!device->mesh->tiles[tile_id].free) 
+				return HB_MC_FAIL;
+		}
 	}
+
 	return HB_MC_SUCCESS;
 }
 
-/*!
- * Initializes Manycore tiles so that they may run kernels.
- * @param fd userspace file descriptor
- * @param eva_id specifies what the EVA-NPA mapping is.
- * @param tiles an array of tile_t structs to initialize.
- * @param num_tiles the number of tiles to initialize.
+
+
+
+/**
+ * Takes in a device and tile group and an origin, initializes tile group and sends packets to all tiles in tile group to set their symbols. 
+ * @param[in]  device        Pointer to device
+ * @param[in]  tg            Pointer to tile group
+ * @param[in]  origin        Origin coordinates of tile group
  * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
  */
-int hb_mc_device_finish (uint8_t fd, eva_id_t eva_id, tile_t *tiles, uint32_t num_tiles) {
-	if (eva_id != 0) {
-		return HB_MC_FAIL; /* eva_id not supported */
-	} 
+static int hb_mc_tile_group_initialize_tiles (hb_mc_device_t *device, hb_mc_tile_group_t *tg, hb_mc_coordinate_t origin) { 
 
-	else if (!mem_manager[eva_id])
-		return HB_MC_SUCCESS; /* there is no memory manager to deinitialize */
-	
-	delete(mem_manager[eva_id]);
-	
-	for (int i = 0; i < num_tiles; i++) { /* freeze tiles */
-		hb_mc_tile_freeze_dep(fd, tiles[i].x, tiles[i].y);
-	}
+	int error;
+	hb_mc_coordinate_t tile_coord;
+	hb_mc_idx_t tile_id;
 
-	int error = hb_mc_fifo_finish(fd);
-	if (error != HB_MC_SUCCESS) {
-		fprintf(stderr, "hb_mc_device_finish() --> hb_mc_host_finish(): failed to terminate host.\n");
+	tg->origin = origin;
+
+	error = hb_mc_origin_eva_map_init (tg->map, origin); 
+	if (error != HB_MC_SUCCESS) { 
+		bsg_pr_err("%s: failed to initialize grid %d tile group (%d,%d) eva map origin.\n", __func__, tg->grid_id, hb_mc_coordinate_get_x(tg->id), hb_mc_coordinate_get_y(tg->id));
 		return HB_MC_FAIL;
 	}
 
+	for (hb_mc_idx_t x = hb_mc_coordinate_get_x(origin); x < hb_mc_coordinate_get_x(origin) + hb_mc_dimension_get_x(tg->dim); x++){
+		for (hb_mc_idx_t y = hb_mc_coordinate_get_y(origin); y < hb_mc_coordinate_get_y(origin) + hb_mc_dimension_get_y(tg->dim); y++){
+			tile_id = hb_mc_get_tile_id (device->mesh->origin, device->mesh->dim, hb_mc_coordinate(x, y));
+
+			device->mesh->tiles[tile_id].origin = origin;
+			device->mesh->tiles[tile_id].tile_group_id = tg->id;
+			device->mesh->tiles[tile_id].free = 0;
+
+			hb_mc_coordinate_t coord_val = hb_mc_coordinate((x - hb_mc_coordinate_get_x(origin)), (y - hb_mc_coordinate_get_y(origin)));
+
+
+			error = hb_mc_tile_set_origin_registers(device->mc, &(device->mesh->tiles[tile_id].coord), &(device->mesh->tiles[tile_id].origin));
+			if (error != HB_MC_SUCCESS) { 
+				 bsg_pr_err("%s: failed to set tile (%d,%d) tile group origin registers CSR_TGO_X/Y.\n", __func__, hb_mc_coordinate_get_x(device->mesh->tiles[tile_id].coord), hb_mc_coordinate_get_y(device->mesh->tiles[tile_id].coord));
+				return HB_MC_FAIL;
```

Comment:
You might want to consider returning the value of error, instead of HB_MC_FAIL.

Reply:
Great pointer! I replaced them in the library and the regression tests. thanks!